model-optimizer/extensions/front/tf/KerasRNNTransformation.py

```diff
@@ -0,0 +1,268 @@
+
+        tensor_list_set_item_node = internal_match['concatenation']
+        tensor_list_set_item_node_name = tensor_list_set_item_node.soft_get('name', tensor_list_set_item_node.id)
+        list_result_node = internal_match['concatenation_result']
+
+        # replace TensorListSetItem with Unsqueeze and use axis attribute for corresponding Result node
+        # to concatenate results from different iterations
+        unsqueeze_list_element = create_op_with_const_inputs(body_graph, Unsqueeze, {1: int64_array(0)},
+                                                             {'name': 'TensorListSetItemUnsqueeze'})
+        tensor_list_set_item_node.in_port(2).get_connection().set_destination(unsqueeze_list_element.in_port(0))
+        tensor_list_set_item_node.out_port(0).get_connection().set_source(unsqueeze_list_element.out_port(0))
+        rename_nodes([(tensor_list_set_item_node, tensor_list_set_item_node_name + '/AbandonedName'),
+                      (unsqueeze_list_element, tensor_list_set_item_node_name)])
+        list_result_node_layer_id = list_result_node.internal_layer_id
+        Loop.update_port_map_value_ext(loop_node.output_port_map, 'internal_layer_id', list_result_node_layer_id,
+                                       'axis', 0)
+
+        # remove TensorListStack to by-pass the node since the result from the Loop node is already concatenated
+        stack_node.out_port(0).get_connection().set_source(stack_node.in_port(0).get_connection().get_source())
+
+        # disconnect ListReserve node because it is no longer needed for Loop
+        list_reserve_node.out_port(0).disconnect()
+
+        # connect a number of iterations with trip count that can be received from the second input of ListReserve
+        # create a constant network with True value for execution_condition so that IE can ignore execution condition
+        # and perform trip_counts iterations. This approach with known trip count value allows to avoid dynamism.
+        loop_node.in_port(1).disconnect()
+        list_reserve_node.in_port(1).get_source().connect(loop_node.in_port(1))
+        for record in loop_node.output_port_map:
+            if 'purpose' in record and record['purpose'] == 'execution_condition':
+                exec_cond_layer_id = record['internal_layer_id']
+                exec_cond_node = Loop.get_body_node_by_internal_id(loop_node, exec_cond_layer_id)
+                const_true = Const(body_graph, {'value': np.array(True, dtype=np.bool)}).create_node()
+                exec_cond_node.in_port(0).get_connection().set_source(const_true.out_port(0))
+
+    def replace_sub_graph(self, graph: Graph, external_match: dict):
+        loop_node = external_match['while']
+        body_graph = loop_node['body']
+        body_pattern = KerasRNNOutputConcatenation.get_body_pattern()
+
+        internal_matches = find_subgraph_match_to_pattern(body_graph, body_pattern)
+
+        if len(internal_matches) == 1:
```

Comment:
We need to support the case when there are more than one output with data from the Loop. Because currently the transformation modifies the sub-graph related to the iteration number and so the same pattern cannot be applied to the second output with data. For now, let's leave the approach as is but create a ticket to solve this issue. The new approach will be useful to support generic Loops (not RNN-like). And also we will rename the transformation to something more generic like `WhileOutputConcatenation`

Reply:
Indeed, now it is a sort of a limitation that is able to detect and process only one chain of TensorListSetItem->TensorListUnstack. In case of multiple output concatenation in the scope of one While operation only one chain is processed and others left as is. The same problem exists for InputSlicing. I will create a ticket for this.

Reply:
Created a ticket #47280